### PR TITLE
Variable order space: h-(de)refinement transfer operator

### PR DESCRIPTION
**Pull request by psocratis**
This PR, in its current state, implements the transfer operator from a variable order coarse space to an h-(de)refined space.
TODO 
- [x] Transfer op ("matrix-free")
- [x] Transfer op as a `SparseMatrix`
- [x]  Derefinement  Operator

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2854](https://github.com/mfem/mfem/pull/2854) | @psocratis | @tzanio | @dylan-copeland + @kmittal2 | 4/7/24 | 6/5/24 | 6/11/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by psocratis** [comment]
@kmittal2 , I think this implements what you wanted. Please take a look and we can push it forward if it looks good.

File: fem/fespace.cpp [hunk 1]
```diff
@@ -1523,6 +1523,69 @@ SparseMatrix *FiniteElementSpace::RefinementMatrix_main(
...
+               P->SetRow(r, coarse_vdofs, row);
+               mark[m] = 1;
+            }
+         }
+      }
+   }
+
+   // Not sure if this check makes sense in the variable order case
+   // MFEM_ASSERT(mark.Sum() == P->Height(), "Not all rows of P set.");
+   if (elem_geoms.Size() != 1) { P->Finalize(); }
```
Comment (dylan-copeland) [inline fem/fespace.cpp:1585]: A comment would be helpful on the meaning of `elem_geoms.Size() != 1` and why `P` is finalized only in that case.

File: fem/fespace.cpp [hunk 2]
```diff
@@ -1674,11 +1748,26 @@ void FiniteElementSpace::RefinementOperator::Mult(const Vector &x,
 
    Vector subY, subX;
 
+   DenseMatrix eP;
+   IsoparametricTransformation isotr;
+   const FiniteElement *fe = nullptr;
```
Comment (dylan-copeland) [inline fem/fespace.cpp:1753]: Define inside the loop?

File: fem/fespace.cpp [hunk 3]
```diff
@@ -2123,14 +2232,34 @@ SparseMatrix* FiniteElementSpace::DerefinementMatrix(int old_ndofs,
...
+         fe = GetFE(emb.parent);
+         const DenseTensor &pmats = dtrans.point_matrices[geom];
+         const int ldof = fe->GetDof();
+
+         IsoparametricTransformation isotr;
+         isotr.SetIdentityTransformation(geom);
+
+         localRVO.SetSize(ldof, ldof);
+         isotr.SetPointMat(pmats(emb.matrix));
+         // Local restriction is size ldofxldof assuming that the parent and
```
Comment (dylan-copeland) [inline fem/fespace.cpp:2253]: Does this assumption always hold? Can't the orders be changed arbitrarily?

File: fem/fespace.cpp [hunk 4]
```diff
@@ -2156,11 +2285,11 @@ SparseMatrix* FiniteElementSpace::DerefinementMatrix(int old_ndofs,
       }
    }
 
-   if (!is_dg)
-   {
-      MFEM_VERIFY(num_marked == R->Height(),
-                  "internal error: not all rows of R were set.");
-   }
+   // if (!is_dg)
```
Comment (dylan-copeland) [inline fem/fespace.cpp:2288]: Maybe keep this with `&& !IsVariableOrder()`.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by psocratis** [comment]
> There are some CI failures with the latest merge 👇

yes, I am aware, thanks! There is also another bug I discovered. I'll push the fix later today

File: fem/fespace.hpp [hunk 5]
```diff
@@ -469,6 +469,9 @@ class FiniteElementSpace
                                        const Table *coarse_elem_fos,
                                        const DenseTensor localP[]) const;
 
+   SparseMatrix *VariableOrderRefinementMatrix_main(const int coarse_ndofs,
```
Comment (kmittal2) [inline fem/fespace.hpp:472]: Add documentation.

**Review by kmittal2 — APPROVED** [review]
This was much needed, thanks @psocratis.

**Review by dylan-copeland — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: fem/fespace.cpp [hunk 6]
```diff
@@ -1525,6 +1525,67 @@ SparseMatrix *FiniteElementSpace::RefinementMatrix_main(
    return P;
 }
 
+SparseMatrix *FiniteElementSpace::VariableOrderRefinementMatrix_main(
```
Comment (jandrej) [inline fem/fespace.cpp:1528]: Why is this named `_main`? There's no other variant right?

Comment (psocratis) [reply fem/fespace.cpp]: good point! Fixed.

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

File: tests/unit/fem/test_hp_transfer.cpp [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+      fes.GetProlongationMatrix()->Mult(temp,err_gf);
+   }
+
+   // 6a. Check if the restricted GridFunction to the de-refined
+   //     mesh exactly reproduces the polynomial GridFunction
+   REQUIRE(err_gf.Norml2() < 1e-11);
+   delete fec;
+}
+
+}
```
Comment (tzanio) [inline tests/unit/fem/test_hp_transfer.cpp:298]:
```suggestion
}

```

**Pull request merged**